evm/src/bn254.rs

```diff
@@ -0,0 +1,743 @@
+use std::str::FromStr;
+
+use ethereum_types::U256;
+use itertools::Itertools;
+use rand::{thread_rng, Rng};
+
+pub const BN_BASE: U256 = U256([
+    4332616871279656263,
+    10917124144477883021,
+    13281191951274694749,
+    3486998266802970665,
+]);
+
+pub type Fp = U256;
+pub type Fp2 = [U256; 2];
+pub type Fp6 = [Fp2; 3];
+pub type Fp12 = [Fp6; 2];
+
+pub fn fp12_to_vec(f: Fp12) -> Vec<U256> {
+    f.into_iter().flatten().flatten().collect()
+}
+
+pub fn fp12_to_array(f: Fp12) -> [U256; 12] {
+    let [[[f0, f1], [f2, f3], [f4, f5]], [[f6, f7], [f8, f9], [f10, f11]]] = f;
+    [f0, f1, f2, f3, f4, f5, f6, f7, f8, f9, f10, f11]
+}
+
+pub fn vec_to_fp12(xs: Vec<U256>) -> Fp12 {
+    xs.into_iter()
+        .tuples::<(U256, U256)>()
+        .map(|(v1, v2)| [v1, v2])
+        .tuples()
+        .map(|(a1, a2, a3, a4, a5, a6)| [[a1, a2, a3], [a4, a5, a6]])
+        .next()
+        .unwrap()
+}
+
+pub type Curve = [Fp; 2];
+pub type TwistedCurve = [Fp2; 2];
+
+pub fn curve_generator() -> Curve {
+    [U256::one(), U256::from(2)]
+}
+
+pub fn twisted_curve_generator() -> TwistedCurve {
+    [
+        [
+            U256::from_str("0x1800deef121f1e76426a00665e5c4479674322d4f75edadd46debd5cd992f6ed")
+                .unwrap(),
+            U256::from_str("0x198e9393920d483a7260bfb731fb5d25f1aa493335a9e71297e485b7aef312c2")
+                .unwrap(),
+        ],
+        [
+            U256::from_str("0x12c85ea5db8c6deb4aab71808dcb408fe3d1e7690c43d37b4ce6cc0166fa7daa")
+                .unwrap(),
+            U256::from_str("0x90689d0585ff075ec9e99ad690c3395bc4b313370b38ef355acdadcd122975b")
+                .unwrap(),
+        ],
+    ]
+}
+
+const ZERO: Fp = U256([0, 0, 0, 0]);
+
+fn embed_fp2(x: Fp) -> Fp2 {
+    [x, ZERO]
+}
+
+fn embed_fp2_fp6(a: Fp2) -> Fp6 {
+    [a, embed_fp2(ZERO), embed_fp2(ZERO)]
+}
+
+fn embed_fp6(x: Fp) -> Fp6 {
+    embed_fp2_fp6(embed_fp2(x))
+}
+
+fn embed_fp12(x: Fp) -> Fp12 {
+    [embed_fp6(x), embed_fp6(ZERO)]
+}
+
+fn gen_fp() -> Fp {
+    let mut rng = thread_rng();
+    let x64 = rng.gen::<u64>();
+    U256([x64, x64, x64, x64]) % BN_BASE
+}
+
+fn gen_fp2() -> Fp2 {
+    [gen_fp(), gen_fp()]
+}
+
+fn gen_fp6() -> Fp6 {
+    [gen_fp2(), gen_fp2(), gen_fp2()]
+}
+
+pub fn gen_fp12() -> Fp12 {
+    [gen_fp6(), gen_fp6()]
+}
+
+pub fn gen_fp12_sparse() -> Fp12 {
+    sparse_embed(gen_fp(), [gen_fp(), gen_fp()], [gen_fp(), gen_fp()])
+}
```

Comment:
This is a very un-Rusty way to do pseudorandom number generation.

Each of these functions should take an `rng` instead of accessing `thread_rng`. This lets the user decide on:
1. The PRNG algorithm to use.
2. The initial seed.

Even better each of `Fpblah` should implement [`rand::Rng`](https://docs.rs/rand/0.8.5/rand/trait.Rng.html) so then you can just do (e.g.) `rng::gen::<Fp12>()`.